File: hyein/week8/단어변환.py
from collections import deque

def solution(begin, target, words):
    if target not in words:
        return 0

    q = deque()
    q.append((begin, 0))

    while q:
        x, cnt = q.popleft()

        if x == target:
            return cnt

        for word in words:
            diff = 0
            for i in range(len(x)):
                if x[i] != word[i]:
                    diff += 1
            if diff == 1:
                q.append((word, cnt + 1))
    return 0

File: sujeong/dfs:bfs/단어 변환.swift
import Foundation

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    
    if !words.contains(target) {
        return 0
    }
    
    // bfs
    var visit:[Bool] = Array(repeating: false, count: words.count)
    var queue:[(begin:String, cnt:Int)] = []
    
    queue.append((begin, 0))
    
    while (!queue.isEmpty) {
        let here:(begin:String,cnt:Int) = queue.removeFirst()
        if(here.begin == target){
            return here.cnt
        }
        
        for (idx,word) in words.enumerated() {
            if !visit[idx] && isADifference(here.begin, word) {
                queue.append((word, here.cnt+1))
                visit[idx] = true
            }
        }
    }
    return 0
}

func isADifference(_ begin:String, _ next:String) -> Bool {
    let a:[Character] = Array(begin)
    let b:[Character] = Array(next)
    
    var cnt:Int = 0
    
    for i in 0..<a.count {
        if a[i] != b[i] {
            cnt += 1
            if cnt > 1 {
                return false
            }
        }
    }
    
    if cnt == 1 {
        return true
    }
    return false
}
